### PR TITLE
feat(optimizer): improve pull_up_correlated_predicate_agg_rule to optimize TPCH Q17

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Resolve https://github.com/risingwavelabs/risingwave/issues/15247#event-11962866299
- Support `count` aggcall for `pull_up_correlated_predicate_agg_rule` if it can be proved that the row generated by empty input finally being able to be eliminated.
- Actually, the Q17 could be further optimized with the `Window` operator, but it is out of the scope of this PR.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by st1page** [comment]
> Actually, the Q17 could be further optimized with the Window operator, but it is out of the scope of this PR.

I am not sure if it could be better in stream because the streaming OverWindow operator is more expensive

**Comment by chenzl25** [comment]
> > Actually, the Q17 could be further optimized with the Window operator, but it is out of the scope of this PR.
> 
> I am not sure if it could be better in stream because the streaming OverWindow operator is more expensive

It reminds me why our agg executor could emit messages when the barrier arrives, however, window executor can't do it in the same way?

**Review by lmatz — APPROVED** [review]
The improved TPC q17 LGTM! 
link: https://github.com/risingwavelabs/risingwave/issues/14799#issuecomment-1909776077

File: src/frontend/src/optimizer/rule/pull_up_correlated_predicate_agg_rule.rs [hunk 1]
```diff
@@ -145,12 +157,52 @@ impl Rule for PullUpCorrelatedPredicateAggRule {
...
+            // If no null agg, bail out.
+            if null_agg_pos.is_empty() {
+                return None;
+            }
+
+            // Try to prove that the expression is null-rejected by the top filter when not-count-aggs are null.
+            let mut agg_null_bitset = FixedBitSet::with_capacity(agg.base.schema().len());
+            for pos in null_agg_pos {
+                agg_null_bitset.insert(pos);
+            }
```
Comment (fuyufjh) [inline src/frontend/src/optimizer/rule/pull_up_correlated_predicate_agg_rule.rs:184]: Are you looking for [`FixedBitSet::from_iter`](https://docs.rs/fixedbitset/0.4.2/fixedbitset/struct.FixedBitSet.html#impl-FromIterator%3Cusize%3E)?

**Review by fuyufjh — COMMENTED** [review]
LGTM

File: src/frontend/src/optimizer/rule/pull_up_correlated_predicate_agg_rule.rs (hunk 1 above)
Comment (chenzl25) [reply src/frontend/src/optimizer/rule/pull_up_correlated_predicate_agg_rule.rs:184]: I need to ensure the capacity of this bitset is `agg.base.schema().len()`.

**Pull request merged**